lib/node_modules/@stdlib/ndarray/flatten/lib/main.js

```diff
@@ -0,0 +1,100 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2025 The Stdlib Authors.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var isndarrayLike = require( '@stdlib/assert/is-ndarray-like' );
+var isPositiveInteger = require( '@stdlib/assert/is-positive-integer' );
+var getData = require( '@stdlib/ndarray/data-buffer' );
+var getShape = require( '@stdlib/ndarray/shape' );
+var getDtype = require( '@stdlib/ndarray/dtype' );
+var getOrder = require( '@stdlib/ndarray/order' );
+var shape2strides = require( '@stdlib/ndarray/base/shape2strides' );
+var ndarray = require( '@stdlib/ndarray/ctor' );
+var format = require( '@stdlib/string/format' );
+
+
+// MAIN //
+
+/**
+* Returns a flattened copy of an input ndarray.
+*
+* @param {ndarray} x - input ndarray
+* @param {integer} depth - number of dimensions to flatten
+* @throws {TypeError} first argument must be an ndarray-like object
+* @throws {RangeError} first argument must be an ndarray-like object with more than one dimension
+* @throws {TypeError} second argument must be a non-negative integer
+* @throws {RangeError} second argument must be not exceed number of input ndarray dimensions
+* @returns {ndarray} output ndarray
+*
+* @example
+* var array = require( '@stdlib/ndarray/array' );
+* var ndarray2array = require( '@stdlib/ndarray/to-array' );
+*
+* // Create an input ndarray:
+* var x = array( [ [ [ 1.0, 2.0 ] ], [ [ 3.0, 4.0 ] ], [ [ 5.0, 6.0 ] ] ] );;
+* // return <ndarray>
+*
+* // Flatten the input ndarray:
+* var y = flatten( x, 2 );
+* // returns <ndarray>
+*
+* var arr = ndarray2array( y );
+* // returns [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ]
+*/
+function flatten( x, depth ) {
+	var ndims;
+	var xsh;
+	var osh;
+	var d;
+	var s;
+	var i;
+
+	if ( !isndarrayLike( x ) ) {
+		throw new TypeError( format( 'invalid argument. First argument must be an ndarray-like object. Value: `%s`.', x ) );
+	}
+
+	xsh = getShape( x );
+	ndims = xsh.length;
+	if ( ndims < 2 ) {
+		throw new RangeError( format( 'invalid argument. First argument be an ndarray-like object with more than one dimension. Value: `%s`.', ndims ) );
+	}
+	d = ndims;
+	if ( isPositiveInteger( depth ) ) {
+		if ( depth > ndims - 1 ) {
```

Comment:
Why should this be an error condition? Why not just clamp to the number of dimensions? I am not sure how likely this is, but I suppose a user could want to flatten a list of ndarrays of varying ranks and want to flatten them to at most a specified depth. By raising an exception here, you disallow this kind of usage.

Think of this sort of clamping behavior as being similar to `indexOf` and friends in the handling of out-of-bounds values.

Reply:
Similar to https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/flatten-array, `depth` should be considered the "maximum depth" to flatten, rather than a strict edict.

You should also always sanity check against built-in behavior:

```
In [29]: [[1,2,3]].flat()
Out[29]: [ 1, 2, 3 ]

In [30]: [[1,2,3]].flat(1)
Out[30]: [ 1, 2, 3 ]

In [31]: [[1,2,3]].flat(10)
Out[31]: [ 1, 2, 3 ]
```

Notice the behavior in the last example.